extended/src/main/java/io/kubernetes/client/extended/controller/DefaultController.java

```diff
@@ -0,0 +1,295 @@
+                  ready = ready && cacheHasSyncedFunc.get();
+                }
+                return ready;
+              });
+    }
+    return synced;
+  }
+
+  @Override
+  public void run() {
+    if (!preFlightCheck()) {
+      log.error("Controller {} failed pre-run check, exiting..", this.name);
+      return;
+    }
+
+    // spawns worker threads for the controller.
+    CountDownLatch latch = new CountDownLatch(workerCount);
+    for (int i = 0; i < this.workerCount; i++) {
+      final int workerIndex = i;
+      workerThreadPool.scheduleWithFixedDelay(
+          () -> {
+            if (log.isDebugEnabled()) {
+              log.debug("Starting controller {} worker {}..", this.name, workerIndex);
+            }
+            this.worker();
+            latch.countDown();
+            if (log.isDebugEnabled()) {
+              log.debug("Exiting controller {} worker {}..", this.name, workerIndex);
+            }
+          },
+          0,
+          1,
+          TimeUnit.SECONDS);
+    }
+    try {
+      if (log.isDebugEnabled()) {
+        log.debug("Controller {} bootstrapping..", this.name);
+      }
+      latch.await();
+    } catch (Throwable t) {
+      log.error("Controller {} hitting exception.", this.name);
+      log.error("Aborting controller.", t);
+    } finally {
+      log.info("Controller {} exited", this.name);
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    // shutdown work-queue before the thread-pool
+    workQueue.shutDown();
+    workerThreadPool.shutdown();
+  }
+
+  private void worker() {
+    // taking tasks from work-queue in a loop
+    while (!workQueue.isShuttingDown()) {
+      Reconciler.Request request = null;
+      try {
+        request = workQueue.get();
+        // request is expected to be null, when the work-queue is shutting-down.
+        if (request == null) {
+          log.info("Controller {} worker exiting because work-queue has shutdown..");
+          return;
+        }
+        if (log.isDebugEnabled()) {
+          log.debug("Controller {} start reconciling {}..", request);
+        }
+
+        // do reconciliation, invoke user customized logic.
+        Reconciler.Result result = this.reconciler.reconcile(request);
+
+        // checks whether do a re-queue (on failure)
+        if (result.isRequeue()) {
+          if (result.getRequeueAfter() == null) {
+            if (log.isDebugEnabled()) {
+              log.debug("Controller {} reconciling {} failed, requeuing {}..", request);
+            }
+            workQueue.addRateLimited(request);
+          } else {
+            if (log.isDebugEnabled()) {
+              log.debug(
+                  "Controller {} reconciling {} failed, requeuing after {}..",
+                  request,
+                  result.getRequeueAfter());
+            }
+            workQueue.addAfter(request, result.getRequeueAfter());
+          }
+        }
+      } catch (InterruptedException e) {
```

Comment:
What in this code throws `InterruptedException` can we wrap the potential throw more tightly?

Reply:
it `BlockingQueue#get` throwing `InterruptedException` and there we have a testing verifying the behavior when a potential throw happens from Reconciler, see `testControllerStopsWorkingWhenReconcilerAbortsWithRuntimeException `

Reply:
Can we just put the catch around that line of code?

Reply:
done